File: day1/day1.php
<?php

$calories_per_elf = \array_map(
    static fn(string $elf_calories) => \array_sum(\explode("\n", $elf_calories)),
    \explode("\n\n", \file_get_contents(dirname(__FILE__) . '/input.txt'))
);

\rsort($calories_per_elf);

\printf("Calories of elf with most calories is: %d (part 1)\n", \max($calories_per_elf));
\printf("Combined calories of top 3 elves is: %d (part 2)\n", \array_sum(\array_slice($calories_per_elf, 0, 3)));

File: day1/day1.hs
import Data.List

type Calories = Int
type Inventory = [Calories]
type CaloriesPerElf = [[Calories]]
type TotalCaloriesPerElf = [Calories]

loadInventory :: IO Inventory
loadInventory = do
    input <- readFile "input.txt"
    return $ map (\x -> if x /= "" then read x :: Int else 0) $ lines input

groupByElf :: Inventory -> CaloriesPerElf
groupByElf xs = case 0 `elemIndex` xs of
    Just i -> [take i xs] ++ groupByElf (drop (i+1) xs)
    Nothing -> []

sumByElf :: Inventory -> TotalCaloriesPerElf
sumByElf xs = map sum $ groupByElf xs

maxElfCalories :: Inventory -> Calories
maxElfCalories xs = maximum $ sumByElf xs

caloriesOfTop3Elves :: Inventory -> Calories
caloriesOfTop3Elves xs = sum $ take 3 $ reverse $ sort $ sumByElf xs

main = do
    inventory <- loadInventory
    putStrLn ("Elf with most calories has " ++ (show $ maxElfCalories inventory) ++ " calories. (part 1)")
    putStrLn ("Top 3 elves have " ++ (show $ caloriesOfTop3Elves inventory) ++ " calories combined. (part 2)")
